feat(portal): adds portal new recipe command for api recipes

aliasghar98 opened the pull request:
Closes #62 .

## What
Adds a new `apimatic portal:recipe:new` command, which follows a wizard-style flow to add an API Recipe to the build input.

aliasghar98 committed:
feat: replaced portal generation axios call with sdk controller method

aliasghar98 committed:
refactor: removed sdk reference from package.json

aliasghar98 committed:
feat: made sdk part of the repo

aliasghar98 committed:
Merge branch 'dev' into aliasghar98/update-apimatic-sdk-version

aliasghar98 committed:
refactor: improved portal generate command functionality

aliasghar98 committed:
chore: updated launch.json to include config for debugging tests

aliasghar98 committed:
chore: updated sdk after adding problem details to spec

aliasghar98 committed:
chore(deps): updated oclif/test, chai, node, nock and mocha dependencies versions

aliasghar98 committed:
fix: resolved a bug related to file stream not closing properly and updated a util method due to node type depedency update

aliasghar98 committed:
feat(tests): added tests for portal:generate

aliasghar98 committed:
refactor: cleaned up code for portal:generate command, controller and updated references accordingly

aliasghar98 committed:
fix: existing portal prompt shows up alongside generating portal spinner

aliasghar98 committed:
refactor: removed duplicate check for problem details error

aliasghar98 committed:
refactor: removed unnecesssary cast and clean up imports

aliasghar98 committed:
refactor: improved portal generate command by separating layers

aliasghar98 committed:
fix(tests): updated tests to conform with refactored approach and fixed minor issues

aliasghar98 committed:
chore: save current WIP state

aliasghar98 committed:
feat(recipe): added types

aliasghar98 committed:
chore: saving current WIP state

aliasghar98 committed:
feat(recipe): added recipe builder and generator classes

aliasghar98 committed:
chore: save current WIP state

aliasghar98 committed:
refactor: removed unnecessary markdown file

aliasghar98 committed:
feat(recipe): added remaining implementation

aliasghar98 committed:
Merge branch 'dev' into aliasghar98/62-add-portal-new-recipe-command-for-api-recipes

aliasghar98 committed:
refactor(recipe): improved implementation

aliasghar98 committed:
fix(recipe): fixed invalid imports

aliasghar98 committed:
chore(recipe): added a todo for dynamic recipe registration in build config

aliasghar98 committed:
refactor(recipe): improved implementation and sorted out folder structure

aliasghar98 committed:
fix(recipe): fixed incorrect error logging

aliasghar98 committed:
test(recipe): added unit tests

aliasghar98 committed:
fix(recipe): fixed incorrect build file name matching

saeedjamshaid commented on src/commands/portal/new/recipe.ts:
```diff
@@ -0,0 +1,25 @@
+  static override flags = {
+    name: Flags.string({ char: "n", description: "name for the recipe" }),
+    folder: Flags.string({
+      char: "d",
```
It is better to remove the character `d` for the flag as this flag represents `folder` and `d` is more appropriate for `destination`. We can't use `f` character as that usually means force

aliasghar98 replied:
Fixed this.

aliasghar98 committed:
fix(recipe): improved messaging and command flags

aliasghar98 committed:
fix(recipe): updated runtime dependencies and fixed minor messaging issue

aliasghar98 committed:
fix(recipe): removed unnecessary dash appended to recipe name

aliasghar98 committed:
fix: resolves duplicate api recipe issue

aliasghar98 committed:
fix: added missing check for not adding duplicate API Recipe in TOC.

saeedjamshaid commented on src/application/portal/new/recipe-generator.ts:
```diff
@@ -0,0 +1,75 @@
+      script += "stepCallback: async () => {";
+
+      if (step.type === "content") {
+        const contentConfig = step.config as ContentStepConfig;
```
you should move the code for adding `content` and `endpoint` step into separate methods

aliasghar98 replied:
Done.

saeedjamshaid commented on src/actions/portal/new/recipe.ts:
```diff
@@ -0,0 +1,338 @@
+
+    const validationResult = await this.validateBuildDirectoryPath(buildDirectoryPath);
+    if (!validationResult.isSuccess) {
+      this.prompts.logError(getMessageInRedColor(`Unable to generate API Recipe: ${validationResult.error!}`));
```
Return all the failure cases error message in Result to the command and log the errors there.

aliasghar98 replied:
Done.

saeedjamshaid commented on src/actions/portal/new/recipe.ts:
```diff
@@ -0,0 +1,338 @@
+  public async createRecipe(buildDirectoryPath: string, name?: string): Promise<Result<string, string>> {
+    const recipeName = name ?? (await this.prompts.recipeNamePrompt());
+    const recipeFileName = recipeName.split(" ").join("");
+    const recipeBuilder = new PortalRecipeBuilder(recipeName);
```
Rename the `PortalRecipeBuilder` class to `PortalRecipe` and also there is no need to instatiate its object here. You can instead just pass the `recipeName` to `promptUserAndBuildNewRecipe` method and then instatiate the `PortalRecipe` inside it

aliasghar98 replied:
Done.

saeedjamshaid commented on src/actions/portal/new/recipe.ts:
```diff
@@ -0,0 +1,338 @@
+      const stepName = await this.prompts.stepNamePrompt("Step " + idx);
+      switch (stepType) {
+        case StepType.Content: {
+          const contentFilePath = await this.prompts.contentFilePathPrompt();
```
Can the code of individual switch cases for endpoint and content be moved to separate methods?

aliasghar98 replied:
Done.

saeedjamshaid commented on src/actions/portal/new/recipe.ts:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    //TODO: Create a type for the build config and use that here instead of any.
+    const buildConfigResult = await this.parseBuildConfig(buildDirectoryPath);
```
Please re-evaluate the logical order of the steps in this action. The steps that or not dependent on the user should be run first so. Any prompt that gives choice to the user to exit (e.g. overwrite: yes/no) should take place as early as possilble.

aliasghar98 replied:
Done.

saeedjamshaid commented on src/actions/portal/new/recipe.ts:
```diff
@@ -0,0 +1,338 @@
+      }
+    }
+
+    if (!recipeAlreadyExists) {
```
This `if` block should be inside the `addRecipeToToc` method

aliasghar98 replied:
Done.

saeedjamshaid commented on src/actions/portal/new/recipe.ts:
```diff
@@ -0,0 +1,338 @@
+    await this.createMarkdownFile(recipeFileName, contentFolderPath);
+
+    const generatedRecipeScript = await recipeGenerator.createScriptFromRecipe(recipeResult.value!);
+    const generatedRecipeScriptsDirectoryPath = path.join(contentFolderPath, "static", "scripts", "api-recipes");
```
Consider refactoring this code so that `PortalRecipeGenerator` instance is responsible for creating all of the recipe related artifacts e.g. markdown file, script file and updating the build config

aliasghar98 replied:
Done.

aliasghar98 committed:
fix: resolved prefix in build config file issue and refactored overall flow

aliasghar98 committed:
refactor: minor changes

aliasghar98 committed:
Merge branch 'dev' into aliasghar98/62-add-portal-new-recipe-command-for-api-recipes

aliasghar98 committed:
chore: updated package-lock.json

aliasghar98 committed:
refactor: updated messaging, refactored some code and improved handling for build config file path prompt

aliasghar98 committed:
refactor: removed uncaught exception

aliasghar98 committed:
feat(portal): adds portal:new:toc command and other various improvements (#93)

Closes #55, Closes #56, Closes #72, Closes #73.

Co-authored-by: saeedjamshaid <saeed.jamshaid@outlook.com>

semantic-release-bot committed:
chore(release): set `package.json` to 1.1.0-alpha.8 [skip ci]

# [1.1.0-alpha.8](https://github.com/apimatic/apimatic-cli/compare/v1.1.0-alpha.7...v1.1.0-alpha.8) (2025-06-24)

### Features

* **portal:** adds portal:new:toc command and other various improvements ([#93](https://github.com/apimatic/apimatic-cli/issues/93)) ([cfae452](https://github.com/apimatic/apimatic-cli/commit/cfae452b26f2ebf393b22afb82740e5b18c78738)), closes [#55](https://github.com/apimatic/apimatic-cli/issues/55) [#56](https://github.com/apimatic/apimatic-cli/issues/56) [#72](https://github.com/apimatic/apimatic-cli/issues/72) [#73](https://github.com/apimatic/apimatic-cli/issues/73)

aliasghar98 committed:
Merge branch 'alpha' into aliasghar98/62-add-portal-new-recipe-command-for-api-recipes

aliasghar98 committed:
feat(recipe): added selector for endpoint groups and endpoints

aliasghar98 committed:
feat: migrated package.json and tsconfig file to esm

aliasghar98 committed:
chore: fixed eslint files

aliasghar98 committed:
fix: updated cli tests debugging launch configuration

aliasghar98 committed:
chore: updated run.js file to conform with esm

aliasghar98 committed:
chore: updated tsconfig file in tests folder

aliasghar98 committed:
fix: updated imports to follow esm conventions and updated prompts to use new version of oclif

aliasghar98 committed:
fix(tests): updated all tests

aliasghar98 committed:
fix: resolved run.js file incorrect behaviour and updated package.json

aliasghar98 committed:
Merge 'dev' into aliasghar98/62-add-portal-new-recipe-command-for-api-recipes

aliasghar98 committed:
Merge  'aliasghar98/migrate-to-esm-from-commonjs' into aliasghar98/62-add-portal-new-recipe-command-for-api-recipes

aliasghar98 committed:
fix: resolved import issues related to commonjs to esm migration

aliasghar98 committed:
fix: renamed recipe command

aliasghar98 committed:
feat: added markdown editor for content step and removed redundant test files

aliasghar98 committed:
refactor: moved new-recipe.test file to correct directory

aliasghar98 committed:
fix: failing sdl parser test

aliasghar98 committed:
feat(tests): added new unit tests for recipe command

aliasghar98 committed:
fix: renamed folder structure from `api-recipes` to just `recipes`

aliasghar98 committed:
fix: resolved toc group addition issue and other minor fixes

aliasghar98 committed:
fix: improved message functions by using picocolors

aliasghar98 committed:
fix: improved portal quickstart messaging

aliasghar98 committed:
fix: minor issues in portal recipe command

aliasghar98 committed:
fix: updated content

aliasghar98 committed:
fix: minor fixes and messaging improvements

aliasghar98 committed:
fix: minor content improvements

aliasghar98 committed:
fix: minor improvements

aliasghar98 committed:
updated recipe command readme sections and added missing type when creating build directory structure

aliasghar98 committed:
docs(readme): updated readme

aliasghar98 committed:
fix: spacing issues in commands

aliasghar98 committed:
fix: removed unnecessary function call
